File: 2025/day11/python/solution.py
#!/usr/bin/env python3
import sys
from functools import lru_cache

def parse_input(filename):
    """Parse input into a graph (adjacency list)."""
    graph = {}
    with open(filename) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            parts = line.split(': ')
            node = parts[0]
            neighbors = parts[1].split() if len(parts) > 1 else []
            graph[node] = neighbors
    return graph

def part1(graph):
    """Count all paths from 'you' to 'out' using memoization."""
    @lru_cache(maxsize=None)
    def count_paths(node):
        if node == 'out':
            return 1
        if node not in graph:
            return 0
        return sum(count_paths(neighbor) for neighbor in graph[node])

    return count_paths('you')

def part2(graph):
    """
    Count paths from 'svr' to 'out' that visit both 'dac' and 'fft'.
    Since it's a DAG, paths either visit dac before fft, or fft before dac.
    """
    from functools import cache

    def count_paths_to_target(target):
        """Count paths from each node to target."""
        @cache
        def count(node):
            if node == target:
                return 1
            if node not in graph:
                return 0
            return sum(count(neighbor) for neighbor in graph[node])
        return count

    paths_to_out = count_paths_to_target('out')
    paths_to_dac = count_paths_to_target('dac')
    paths_to_fft = count_paths_to_target('fft')

    # Paths that visit dac before fft: svr -> dac -> fft -> out
    dac_before_fft = paths_to_dac('svr') * paths_to_fft('dac') * paths_to_out('fft')

    # Paths that visit fft before dac: svr -> fft -> dac -> out
    fft_before_dac = paths_to_fft('svr') * paths_to_dac('fft') * paths_to_out('dac')

    return dac_before_fft + fft_before_dac

def main():
    input_file = '../input.txt'
    if len(sys.argv) > 1:
        input_file = sys.argv[1]

    graph = parse_input(input_file)

    print(f"Part 1: {part1(graph)}")
    print(f"Part 2: {part2(graph)}")

if __name__ == '__main__':
    main()

File: 2025/day11/ruby/solution.rb
#!/usr/bin/env ruby

def parse_input(filename)
  # Parse input into a graph (adjacency list)
  graph = {}
  File.readlines(filename).each do |line|
    line = line.strip
    next if line.empty?

    parts = line.split(': ')
    node = parts[0]
    neighbors = parts.length > 1 ? parts[1].split(' ') : []
    graph[node] = neighbors
  end
  graph
end

def part1(graph)
  # Count all paths from 'you' to 'out' using memoization
  memo = {}

  count_paths = lambda do |node|
    return memo[node] if memo.key?(node)

    if node == 'out'
      return 1
    end

    if !graph.key?(node)
      return 0
    end

    result = graph[node].sum { |neighbor| count_paths.call(neighbor) }
    memo[node] = result
    result
  end

  count_paths.call('you')
end

def part2(graph)
  # Count paths from 'svr' to 'out' that visit both 'dac' and 'fft'
  # Since it's a DAG, paths either visit dac before fft, or fft before dac

  # Helper to count paths from each node to a target
  count_paths_to_target = lambda do |target|
    memo = {}

    count = lambda do |node|
      return memo[node] if memo.key?(node)

      if node == target
        return 1
      end

      if !graph.key?(node)
        return 0
      end

      result = graph[node].sum { |neighbor| count.call(neighbor) }
      memo[node] = result
      result
    end

    count
  end

  paths_to_out = count_paths_to_target.call('out')
  paths_to_dac = count_paths_to_target.call('dac')
  paths_to_fft = count_paths_to_target.call('fft')

  # Paths that visit dac before fft: svr -> dac -> fft -> out
  dac_before_fft = paths_to_dac.call('svr') * paths_to_fft.call('dac') * paths_to_out.call('fft')

  # Paths that visit fft before dac: svr -> fft -> dac -> out
  fft_before_dac = paths_to_fft.call('svr') * paths_to_dac.call('fft') * paths_to_out.call('dac')

  dac_before_fft + fft_before_dac
end

# Main
input_file = ARGV[0] || '../input.txt'
graph = parse_input(input_file)

puts "Part 1: #{part1(graph)}"
puts "Part 2: #{part2(graph)}"
